README
$Header$

This Works is placed under the terms of the Copyright Less License,
see file COPYRIGHT.CLL.  USE AT OWN RISK, ABSOLUTELY NO WARRANTY.

However as it ekhtml is linked in statically, binaries must
be compliant to the EKTHML license.

Origin: http://www.scylla-charybdis.com/tool.php/tinohtmlparse

Compile:
========

This is currently based on ekhtml, a deadly inperfect HTML parser (for
example it does not parse Comments correctly, like in the following:
<!DTD -- comment 1 -- more DTD -- comment 2 -- again DTD />).  Perhaps
sometimes I come around and write a working version (which then shall
be able to sanitize HTML as well), but for now, we keep it as it is.

First fetch ekhtml from CVS at sourceforge.net:
	cvs -d :pserver:anonymous@ekthml.cvs.sourceforge.net:/cvsroot/ekhtml login
		(empty password)
	cvs -z3 -d :pserver:anonymous@ekhtml.cvs.sourceforge.net:/cvsroot/ekhtml co ekhtml
Compile ekhtml:
	cd ekhtml; ./autogen.sh; make

To compile tinohtmlparse:

Create softlink to the source of ekhtml if you compiled it in another directory:
	ln -s ../somewhere/ekhtml ekhtml
Then type
	make

You can ignore the error that tinolib is missing as tinolib is not
required for this.  If you really want it, grab the distribution of
tinolib and let the softlink point there like this:
		ln -s ../somewhere/tinolib-*/old tinolib
Note that tinolib restricts distributions which bundle it to GPL!


Usage:
======

Usage:
	tinohtmlparse [-r|--raw] [-l|--list] [-o|--old]

--list shows a list of known entities

--raw does not convert these entities to their % representation in
attributes.

--old uses the old output variant.  The new variant separates the
last argument from the previous one with a TAB instead of SPC.

The HTML file is read from stdin and the output is written to stdout.
The parsed lines all look like following template:

TYPE TAG ATTR Q TEXT

The words are separated by SPC or TAB.  The first 4 words are
guaranteed to not contain SPC or TAB ever.  If they are empty it is
guaranteed that no more words or text follows.

- TYPE is a type string (see below)
- TAG usually is the HTML TAG (ekhtml converts this to uppercase)
- ATTR is the attribute name
- Q is a Quote type of the text which follows.
- TEXT is the text and is % escaped such that it is URL compatible

When TYPE is "text" or "comment" then TAG is a number counting the
lines starting with 0, ATTR is an LF flag and Q always is -.

Q can be B for boolean attribs (those without =), N (was not quoted),
' or " (the quote which was used).  There is a form where Quote is two
HEX digits HH, but this never shall show up (it's in case ekhtml send
some unusual quote character).

So you can do

./tinohtmlparse < htmlfile |
./tinohtmlabsurl.sh "BASEURL" |
while read -r type tag name q text
do
	...
done


Output documentation:
=====================

open TAG
close TAG
	Open and closing TAG tags encountered.  TEXT is empty.

attr TAG ATTR Q TEXT
	A named attribute, immediately follows "open".

	TAG is the TAG it belongs to, added for more easy parsing.
	ATTR is the attribute's name.

	The text is URL-escaped with %, that is %xx is the hex
	representation of any unusual character (including SPC).
	For unicode characters there is the representation %uXXXX.

	If HTMLentities (like &amp;) are encountered, they are
	automatically changed into the character representation.
	(if not --raw given.)

text COUNT LF - TEXT
comment COUNT LF - TEXT
	COUNT is the line count.

	LF is either 0 (TEXT does not contain a LF) or 1 (TEXT does
	contain an LF).  Multiple lines are repeated with the line
	count counted up, so there are no complex to parse
	continuation lines.

	In case of the comment form, this is the commented out text.


Notes:
======

People out there partly write deadly HTML code.  tinohtmlparse is
too perfect to handle this - all heuristics are missing.

For example what I already have seen (yes, this are TWO lines):

<img src=" pic
.jpg">

Firefox handles this correctly and loads pic.jpg!  So it silently
trims SPCs from URLs and removes CRs and LFs from within, too.
tinohtmlparse does not do this.  It transforms the URL into
%20pic%0a.jpg, which then probably is not found on the server.

If you need to parse such crap, apply your own heuristics.
tinohtmlparse will never do this for you.

Please also note that before 0.1.4 the SPC was not escaped to
%20, it was output unchanged.  This is fixed now.


-Tino
webmaster@scylla-charybdis.com
$Log$
Revision 1.7  2009-01-28 18:27:29  tino
TAB for separator for last argument changed

Revision 1.6  2007-12-30 17:57:03  tino
Placed under the CLL, also one entity code was fixed (&and;)

Revision 1.5  2007-12-30 17:15:45  tino
SF URL

Revision 1.4  2007-09-16 06:19:05  tino
Output documentation fixed and notes added

Revision 1.3  2007/09/16 06:10:17  tino
corrected

Revision 1.2  2006/06/11 06:57:30  tino
Mainly only documentation corrected

Revision 1.1  2005/02/05 23:07:28  tino
first commit, tinohtmlparse.c is missing "text" aggregation


ChangeLog
2009-07-13  Valentin Hilbig  <tino@zoo.03.softkill.net>

	* tinohtmlabsurl.sh: Relative URL parser fixed, it apparently did
	not work with #anchor and ?query type URLs.

        * tinohtmlabsurl.sh: host now always lowercase (this certainly
	does not support foreign charactersets, but whoever implemented
	that was nuts anyway).

2007-12-30  Valentin Hilbig  <tino@firebird.03.softkill.net>

	* README (Compile): new SF URL for ekthml CVS

2007-09-16    <tino@firebird.03.softkill.net>

	* tinohtmlparse.c (ce): bugfix: % sign removed from % escape (as
	this ususally is a percent escape) and blank added (it's too often
	in buggy URLs).

2007-02-12    <tino@firebird.03.softkill.net>

	* tinohtmlparse.c (main): again fix for incomplete lines.  Now
	only single lines will be fed to ekhtml, in the hope that no more
	buffer overruns will show up again.

2006-06-16    <tino@firebird.03.softkill.net>

	* tinohtmlabsurl.sh (Usage): added and option to send
	non-url-lines to somewhere else (preferably /dev/null).

2006-02-12    <tino@firebird.03.softkill.net>

	* tinohtmlparse.c (cb_start): EKHTML does not parse HTML-entities
	in atrribute strings.  Added this into tinohtmlparse.
	(p_ent): added for HTML-entity parsing
	(cx, ce): added for HTML-entitiy parsing output
	(main): usage added

2005-02-06    <tino@firebird.03.softkill.net>

	* tinohtmlparse.c (main): Ugly hack added to only feed full lines
	to ekhtml, as else cb_data might be called with incomplete lines.
	This is not the right way to do it.  It shall buffer the last line
	in cb_data, and flush the data to stdout as soon something else
	comes up.  However this is fairly complicated.  This way is quick
	and dirty, however the line numbers might get mixed up (restart at
	0 suddenly).  I can live with it.


tinohtmlabsurl.sh
#!/bin/sh
#
# $Header$
#
# THIS IS NOT PERFECT,
# it does not process a lot of URI attributes for now,
# like applet_codebase, applet_code etc.
#
# This Works is placed under the terms of the Copyright Less License,
# see file COPYRIGHT.CLL.  USE AT OWN RISK, ABSOLUTELY NO WARRANTY.
#
# $Log$
# Revision 1.7  2009-08-03 12:41:26  tino
# Lowercase
#
# Revision 1.6  2009-07-12 22:19:44  tino
# Url-Parser fixed in tinohtmlabsurl.sh
#
# Revision 1.5  2007-12-30 17:57:03  tino
# Placed under the CLL, also one entity code was fixed (&and;)
#
# Revision 1.4  2007-02-12 07:00:02  tino
# Commit for dist, see ChangeLog
#
# Revision 1.3  2006/06/16 20:35:41  tino
# Jump now name anchor, which can be understood better.
# Capability to send all non-URL-lines elsewhere, if you concentrate on URLs
#
# Revision 1.2  2005/02/06 00:17:06  tino
# Only full lines are fed to the parser to make output more easy to parse.
#
# Revision 1.1  2005/02/05 23:07:28  tino
# first commit, tinohtmlparse.c is missing "text" aggregation

if [ 1 != "$#" -a 2 != "$#" ]
then
	echo "Usage: tinohtmlparse | `basename "$0"` BASEURL [/dev/stdout]
	Second argument gives file for non-URL lines" >&2
	exit 1
fi

awk -v BASE="$1" -v NOURLOUT="${2:-/dev/stdout}" '
function shift(x)
{
  sub(/^[^[:space:]]*[[:space:]]/,"",x);
  return x;
}

# http://user:pass@hostname/path/to/index.html?var=data#text
# -type-!!-user---!!-host-!!-path--!!-file---!!-query-!!-anchor-
#
# relative URLs: path does not start with /
# Special characters:
# # starts the anchor anywhere in the URL
# ? starts the query part anywhere in the URL
# @ must preceede host, which must precede /
#
# It does not assume directory structure:
# .. becomes path "" name ".."
# ../.. becomes path "../" name ".."
function parseuri(u, b, c)
{
  delete parsed

  # fetch #anchor
  b	= u;
  gsub(/^[^#]*/, "", b);
  gsub(/#.*$/,"",u);
  parsed["anchor"]=b

  # Fetch ?query
  b	= u;
  gsub(/^[^?]*/, "", b);
  gsub(/\?.*$/,"",u);
  parsed["query"]=b

  # Fetch schema://
  b	= "";
  if (match(u,/^[a-z]*:\/?\/?/))
    {
      b	= substr(u,RSTART,RLENGTH);
      u	= substr(u,RSTART+RLENGTH);
    }
  parsed["type"]=b

  # Fetch host if there is a schema
  c	= ""
  if (b!="")
    {
      b	= u
      gsub(/\/.*$/,"",b);
      gsub(/^[^/]*/,"",u);
      # check user:password@
      if (match(b,/@/))
        {
          c	= substr(b,1,RSTART);
          b	= substr(b,RSTART+1);
        }
    }
  parsed["user"]=c;
  parsed["host"]=b;

  # Now left is path/file
  b	= u
  gsub(/^.*\//,"",b);
  parsed["file"]=b;

  gsub(/[^/]*$/,"",u);
  parsed["path"]=u
}

function dump(t,a,s)
{
  s	= sprintf("# %s", t);

  a="type";  s=s sprintf(" %s=\"%s\"", a, parsed[a]);
  a="user";  s=s sprintf(" %s=\"%s\"", a, parsed[a]);
  a="host";  s=s sprintf(" %s=\"%s\"", a, parsed[a]);
  a="path";  s=s sprintf(" %s=\"%s\"", a, parsed[a]);
  a="file";  s=s sprintf(" %s=\"%s\"", a, parsed[a]);
  a="query"; s=s sprintf(" %s=\"%s\"", a, parsed[a]);
  a="anchor";  s=s sprintf(" %s=\"%s\"", a, parsed[a]);

  print s >NOURLOUT
}

function setbase(b)
{
  parseuri(b)

  # Make sure, BASE path starts and ends with /
  # (the latter is the same as base["path"]=="")
  if (parsed["path"]!~/^\// || parsed["path"]!~/\/$/)
    parsed["path"]="/";

  dump("base");

  basetype	= parsed["type"]
  basehost	= parsed["host"]
  basepath	= parsed["path"]
  basefile	= parsed["file"]
  basequery	= parsed["query"]
}

# Make a full URI from a relative one
# The idea is
function makefull(u,p)
{
  parseuri(u)

  dump("1");

  # Bugfix:
  # We have two options here:
  # base: http://example.com/dir/file?query#anchor
  # uri=: ?query
  # Does the result have a file?
  # Browsers think 'yes'
  # This now shall parse #anchor type destinations, too.
  if (parsed["type"]=="" && parsed["host"]=="" && parsed["path"]=="" && parsed["file"]=="" && parsed["query"]=="")
    parsed["query"]=basequery;
  if (parsed["type"]=="" && parsed["host"]=="" && parsed["path"]=="" && parsed["file"]=="")
    parsed["file"]=basefile;

  # Take over type, host and path from BASE URI
  if (parsed["type"]=="")
    parsed["type"]=basetype;
  # Well, what to do when switching from http: to https:?
  # news:whatever has no host!
  if (parsed["host"]=="" && parsed["type"]==basetype)
    parsed["host"]=basehost;
  # This apparently only works when there was no host
  if (parsed["path"]!~/^\//)
    parsed["path"] = basepath parsed["path"]

  dump("2");

  return tolower(parsed["type"]) parsed["user"] tolower(parsed["host"]) parsed["path"] parsed["file"] parsed["query"] parsed["anchor"];
}

BEGIN	{
	tag["base","href"]=1

	# I am conservative
	# Do not touch background,href,src I do not understand.
	tag["a","href"]=2
	tag["area","href"]=2
	tag["bgsound","src"]=2
	tag["body","background"]=2
	tag["embed","src"]=2
	tag["form","action"]=2
	tag["frame","src"]=2
	tag["iframe","src"]=2
	tag["input","src"]=2
	tag["img","src"]=2
	tag["layer","background"]=2
	tag["layer","src"]=2
	tag["link","href"]=2
	tag["script","src"]=2
	tag["table","background"]=2
	tag["td","background"]=2
	tag["th","background"]=2
	setbase(BASE)
	IGNORECASE=1
	}

$1=="attr" && (t=tag[tolower($2),tolower($3)])!=0	{
	uri=shift(shift(shift(shift($0))))
	if (t==1)
		setbase(uri)
	else
		{
		n=makefull(uri)
		if (uri!=n)
		  {
		    print "# " $0 > NOURLOUT
		    print $1 " " $2 " " $3 " " $4 " " n
		    next
		  }
		print
		next
		}
	}

	{ print >NOURLOUT  }
'


tinohtmlparse.c
/* $Header$
 *
 * This Works is placed under the terms of the Copyright Less License,
 * see file COPYRIGHT.CLL.  USE AT OWN RISK, ABSOLUTELY NO WARRANTY.
 *
 * NOTE THAT ABOVE ONLY APPLIES TO THE CODE WITHIN THIS FILE, YOU MUST
 * HONOR THE EKTHML LICENSE IF YOU USE EKHTML!
 *
 * Written from scratch while reading the demo code tester.c from
 * ekhtml.
 *
 * $Log$
 * Revision 1.9  2009-01-28 19:29:31  tino
 * Forgot TAB in meta
 *
 * Revision 1.8  2009-01-28 18:27:29  tino
 * TAB for separator for last argument changed
 *
 * Revision 1.7  2007-12-30 17:57:03  tino
 * Placed under the CLL, also one entity code was fixed (&and;)
 *
 * Revision 1.6  2007-09-16 06:04:49  tino
 * Percent escape handling corrected
 *
 * Revision 1.5  2007/02/12 07:00:02  tino
 * Commit for dist, see ChangeLog
 *
 * Revision 1.4  2006/06/11 06:57:30  tino
 * Mainly only documentation corrected
 *
 * Revision 1.3  2006/02/12 03:35:41  tino
 * new dist for version with html entitiy parsing
 *
 * Revision 1.2  2005/02/06 00:17:06  tino
 * Only full lines are fed to the parser to make output more easy to parse.
 *
 * Revision 1.1  2005/02/05 23:07:28  tino
 * first commit, tinohtmlparse.c is missing "text" aggregation
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <ctype.h>

#include "ekhtml/include/ekhtml.h"

#include "tinohtmlparse_version.h"

#include "tino_html_entities.h"

static int	raw_mode, old_mode;
 
static void
spc(void)
{ 
  putchar(' ');
}

static void
tab(void)
{ 
  putchar(old_mode ? ' ' : '\t');
}

static void
lf(void)
{ 
  putchar('\n');
}

static void
co(char c)
{
  putchar(c);
  if (c=='\n')
    putchar('+');
}

static void
cp(char c)
{
   putchar(isprint(c) && !isspace(c) ? c : '_');
}

static void
cx(int c)
{
  co("0123456789abcdef"[((unsigned)c)&0xf]);
}

static void
ce(int c)
{
  if (c<=32 || c>126) /* || c=='%') */
    {
      co('%');
      cx(c>>4);
      cx(c);
    }
  else
    co(c);
}


/**********************************************************************/

static void
p_b(const char *s, ...)
{
  va_list list;

  va_start(list, s);
  vprintf(s, list);
  va_end(list);
}

static void
p_p(ekhtml_string_t *s)
{
  int	i;

  for (i=0; i<s->len; i++)
    cp(s->str[i]);
}

static void
p_s(ekhtml_string_t *s)
{
  spc();
  p_p(s);
}

static void
p_e(void)
{
  lf();
}

static void
p_2(const char *typ, ekhtml_string_t *tag)
{
  p_b(typ);
  tab();
  p_p(tag);
  p_e();
}

static void
p_t(ekhtml_string_t *s)
{
  int	i, m;

  m	= s->len;
  if (m && s->str[m-1]=='\n')
    m--;
  for (i=0; i<m; i++)
    co(s->str[i]);
}

/* Print string,
 * but parse html_entities
 *
 * Well, I assume that it's a link.
 * So UTF8 will be transformed into % notation.
 */
static void
p_ent(ekhtml_string_t *s)
{
  int	i, m;

  m	= s->len;
  for (i=0; i<m; )
    {
      if (!raw_mode && s->str[i]=='&')
	{
	  int	len;
	  int	u;

	  len	= m-i;
	  u	= tino_html_entity_check(s->str+i, &len);
	  if (u>=0)
	    {
	      if (u>0xff)
		{
		  co('%');
		  co('u');
		  cx(u>>12);
		  cx(u>>8);
		  cx(u>>4);
		  cx(u);
		}
	      else
		ce(u);
	      i	+= len;
	      continue;
	    }
	}
      ce(s->str[i++]);
    }
}

/* Prefixed Message
 */
static void
p_m(const char *prefix, ekhtml_string_t *s)
{
  int	i;
  int	l;

  for (l=0, i=0; i<s->len; l++)
    {
      int	j;

      printf("%s %d ", prefix, l);
      for (j=i;;j++)
	if (j>=s->len)
	  {
	    printf("0 ");
	    break;
	  }
	else if (s->str[j]=='\n')
	  {
	    printf("1 ");
	    break;
	  }
      printf("-");
      tab();
      while (i<j)
        co(s->str[i++]);
      lf();
      i++;
    }
}

/**********************************************************************/

static void
cb_start(void *x, ekhtml_string_t *tag, ekhtml_attr_t *att)
{
  ekhtml_attr_t *attr;

  p_2("open", tag);

  for (attr=att; attr; attr=attr->next)
    {
      p_b("attr");
      p_s(tag);
      p_s(&attr->name);
      spc();
      if (attr->isBoolean)
        {
	  p_b("B");
	}
      else if (attr->quoteChar)
	{
	  switch (attr->quoteChar)
	    {
	    case '"':
	    case '\'':
	      p_b("%c", attr->quoteChar);
	      break;

	    default:
	      p_b("%02x", (unsigned)(unsigned char)attr->quoteChar);
	      break;
	    }
	}
      else
	{
	  p_b("N");
	}
      tab();
      p_ent(&attr->val);
      p_e();
    }
}

static void
cb_end(void *x, ekhtml_string_t *tag)
{
  p_2("close", tag);
}

static void
cb_comment(void *x, ekhtml_string_t *comment)
{
  p_m("comment", comment);
}

/* ekHTML does not call this with full lines.  Instead it calls this
 * with just the data it has so far.  This is correct, but in our case
 * I don't want to have partial lines in normal situations.
 * This is now "fixed" by feeding full lines into the parser.
 */
static void
cb_data(void *x, ekhtml_string_t *data)
{
  p_m("text", data);
}

/**********************************************************************/

int
main(int argc, char **argv)
{
  ekhtml_parser_t	*p;
  ekhtml_string_t	s;
  char			buf[BUFSIZ*10];
  int			fill;
  int			i;

  raw_mode	= 0;
  for (i=1; i<argc; i++)
    {
      if (!strcmp(argv[1], "-r") || !strcmp(argv[1], "--raw"))
	{
	  raw_mode	= !raw_mode;
	  continue;
	}
      if (!strcmp(argv[1], "-l") || !strcmp(argv[1], "--list"))
	{
	  struct tino_html_entities	*p;

	  for (p=tino_html_entities; p->unicode; p++)
	    printf("%04x\t%5d\t%s\n", p->unicode, p->unicode, p->entity);
	  return 0;
	}
      if (!strcmp(argv[1], "-o") || !strcmp(argv[1], "--old"))
	{
	  old_mode	= !old_mode;
	  continue;
	}
      fprintf(stderr,
	      "Usage: %s [options] < HTMLFILE > parsed_output\n"
	      "\t\tVersion " TINOHTMLPARSE_VERSION " compiled " __DATE__ "\n"
	      "\t--raw\t(also -r) Do not interpret htmlentities\n"
	      "\t\tIf you depend on the broken values, give --raw\n"
	      "\t\tElse tinohtmlparse tries to parse things for HTML entities\n"
	      "\t\tand transfor something clever, which probably is unwanted\n"
	      "\t--list\t(also -l) List known htmlentities (without & and ;)\n"
	      "\t\tNote that &#n; and &#xX; also is known but not listed.\n"
	      "\t--old\t(also -o) Use old SPC instead of TAB to separate last arg\n"
	      , argv[0]
	      );
      return 1;
    }
  p = ekhtml_parser_new(NULL);
    
  ekhtml_parser_datacb_set(p, cb_data);
  ekhtml_parser_commentcb_set(p, cb_comment);
  ekhtml_parser_startcb_add(p, NULL, cb_start);
  ekhtml_parser_endcb_add(p, NULL, cb_end);

  fill	= 0;
  for (;;)
    {
      int	n, i, k;

      if (fflush(stdout) || ferror(stdin) || feof(stdout) || ferror(stdout))
        return 1;

      n		= fread(buf+fill, 1, sizeof buf-fill, stdin);
      fill	+= n;

      /* Ugly fix:
       * Only feed full lines to ekhtml,
       * such that cb_data only gets full lines.
       */
      k		= fill;
      for (i=0; i<fill; i++)
	if (buf[i]=='\n')
	  {
	    k	= i+1;
	    break;
	  }

      s.str	= buf;
      s.len	= k;
      ekhtml_parser_feed(p, &s);

      fill	-= k;
      if (fill)
	memmove(buf, buf+k, fill);
      else if (!n)
	break;

#if 1
      ekhtml_parser_flush(p, 0);
#endif
    }
  ekhtml_parser_flush(p, 1);
  ekhtml_parser_destroy(p);
  return 0;
}
